volatility/framework/plugins/windows/dumpfiles.py

```diff
@@ -0,0 +1,268 @@
+                                        description="Dump a single _FILE_OBJECT at this physical address",
+                                        optional=True),
+            requirements.VersionRequirement(name='pslist', component=pslist.PsList, version=(2, 0, 0)),
+            requirements.VersionRequirement(name='handles', component=handles.Handles, version=(1, 0, 0))
+        ]
+
+    @classmethod
+    def dump_file_producer(cls,
+                           file_object: interfaces.objects.ObjectInterface,
+                           memory_object: interfaces.objects.ObjectInterface,
+                           open_method: Type[interfaces.plugins.FileHandlerInterface],
+                           layer: interfaces.layers.DataLayerInterface,
+                           desired_file_name: str) -> Optional[interfaces.plugins.FileHandlerInterface]:
+        """Produce a file from the memory object's get_available_pages() interface.
+
+        :param file_object: the parent _FILE_OBJECT
+        :param memory_object: the _CONTROL_AREA or _SHARED_CACHE_MAP
+        :param open_method: class for constructing output files
+        :param layer: the memory layer to read from
+        :param desired_file_name: name of the output file
+        :return: result status
+        """
+        filedata = open_method(desired_file_name)
+        try:
+            # Description of these variables:
+            #   memoffset: offset in the specified layer where the page begins
+            #   fileoffset: write to this offset in the destination file
+            #   datasize: size of the page
+
+            # track number of bytes written so we don't write empty files to disk
+            bytes_written = 0
+            for memoffset, fileoffset, datasize in memory_object.get_available_pages():
+                data = layer.read(memoffset, datasize, pad = True)
+                bytes_written += len(data)
+                filedata.seek(fileoffset)
+                filedata.write(data)
+
+            if not bytes_written:
+                vollog.debug("No data is cached for the file at {0:#x}".format(file_object.vol.offset))
+                return None
+            else:
+                vollog.debug("Stored {}".format(filedata.preferred_filename))
+                return filedata
+        except exceptions.InvalidAddressException:
+            vollog.debug("Unable to dump file at {0:#x}".format(
+                file_object.vol.offset))
+            return None
+
+    @classmethod
+    def process_file_object(cls, 
+                            context: interfaces.context.ContextInterface, 
+                            primary_layer_name: str,
+                            open_method: Type[interfaces.plugins.FileHandlerInterface],
+                            file_obj: interfaces.objects.ObjectInterface) -> Tuple:
+        """Given a FILE_OBJECT, dump data to separate files for each of the three file caches.
```

Comment:
Awesome to see such good documentation!  5:D  The only thing I'll say is that we do support [Napolean notation](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/), if you ever feel it's not as readable as it could be... 5;)